### DISCUSSION TITLE
How to detect if response is cached or not?

### DISCUSSION THREAD

**Discussion by Onur169** [artilleryio/artillery#1137] [Artillery]
I want to determine the size of the cache if it is active. If this is not active and the response has to be requested again, then I would like to read out the "Content-Length" header. But here I get an "ECONNRESET" error. How do I solve the problem? thanks in advance

**Comment by hassy** [comment]
hi @Onur169 👋 

`ECONNRESET` means that the server closed the connection unexpectedly. This usually happens when the server is becoming overloaded.

You can tell if the response is cached by inspecting response headers (which header can depend on the caching layer you use). You can do this with an `afterResponse` hook. If the header is not set, you can then set a scenario variable (called `notCached` for example), and then use [a conditional request](https://artillery.io/docs/guides/guides/http-reference.html#Conditional-Requests) to make another request if needed. That request can have its own `afterResponse` hook to read the `Content-Lengith` header. You can then log that value, or track it with a [custom metric](https://artillery.io/docs/guides/guides/extending.html#Tracking-custom-metrics)

**Reply by Onur169** [reply]
Thanks for your answer. I've implemented Etags in my API. For this Artillery would have to note the value of the Etag hash and with the next request, which repeats itself, it should send a request with an "If-None-Match" header. I want to detect if the status code is 304 or 200. That will not happen, it's always 200. Do I have to implement the logic regarding the "hash management" myself?

ECONNRESET problem: Unfortunately, that happens when I set the content-length header myself. After that, the entire test does not work. What I noticed is that some API calls remove the content length from the header again when Transfer-Encoding: chunked is active. Maybe this is a bug?

**Reply by hassy** [reply]
Yes, you'd have to implement the logic yourself. Happy to help out to get it working if you share your initial code!

`Content-Length` is omitted when `Transfer-Encoding` is set to `chunked`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#directives

Could the server be dropping the connection because `Content-Length` is not set correctly?
